File: java/example/src/main/java/TimeoutTest.java
import io.antidrift.zeromcp.*;

public class TimeoutTest {
    public static void main(String[] args) {
        var server = new ZeroMcp();

        server.tool("hello", Tool.builder()
            .description("Fast tool")
            .input(Input.required("name", "string"))
            .execute((a, ctx) -> "Hello, " + a.get("name") + "!")
            .build());

        server.tool("slow", Tool.builder()
            .description("Tool that takes 3 seconds")
            .permissions(new Permissions(
                Permissions.NetworkPermission.unset(),
                Permissions.FsPermission.NONE,
                false,
                2000
            ))
            .execute((a, ctx) -> {
                Thread.sleep(3000);
                return java.util.Map.of("status", "ok");
            })
            .build());

        server.serve();
    }
}

File: rust/examples/timeout_test.rs
use serde_json::Value;
use zeromcp::{Ctx, Input, Permissions, Server, Tool};

#[tokio::main]
async fn main() {
    let mut server = Server::new();

    server.tool(
        "hello",
        Tool {
            description: "Fast tool".to_string(),
            input: Input::new().required("name", "string"),
            permissions: Permissions::default(),
            execute: Box::new(|args: Value, _ctx: Ctx| {
                Box::pin(async move {
                    let name = args["name"].as_str().unwrap_or("world");
                    Ok(Value::String(format!("Hello, {name}!")))
                })
            }),
        },
    );

    server.tool(
        "slow",
        Tool {
            description: "Tool that takes 3 seconds".to_string(),
            input: Input::new(),
            permissions: Permissions {
                execute_timeout: Some(2000),
                ..Default::default()
            },
            execute: Box::new(|_args: Value, _ctx: Ctx| {
                Box::pin(async move {
                    tokio::time::sleep(std::time::Duration::from_secs(3)).await;
                    Ok(serde_json::json!({"status": "ok"}))
                })
            }),
        },
    );

    server.serve().await;
}
